bot/cogs/verification.py

```diff
@@ -1,19 +1,46 @@
+import asyncio
 import logging
+import typing as t
 from contextlib import suppress
+from datetime import datetime, timedelta
 
-from discord import Colour, Forbidden, Message, NotFound, Object
-from discord.ext.commands import Cog, Context, command
+import discord
+from discord.ext import tasks
+from discord.ext.commands import Cog, Context, command, group
+from discord.utils import snowflake_time
 
 from bot import constants
 from bot.bot import Bot
 from bot.cogs.moderation import ModLog
-from bot.decorators import in_whitelist, without_role
+from bot.decorators import in_whitelist, with_role, without_role
 from bot.utils.checks import InWhitelistCheckFailure, without_role_check
+from bot.utils.redis_cache import RedisCache
 
 log = logging.getLogger(__name__)
 
-WELCOME_MESSAGE = f"""
-Hello! Welcome to the server, and thanks for verifying yourself!
+UNVERIFIED_AFTER = 3  # Amount of days after which non-Developers receive the @Unverified role
+KICKED_AFTER = 30  # Amount of days after which non-Developers get kicked from the guild
+
+# Number in range [0, 1] determining the percentage of unverified users that are safe
+# to be kicked from the guild in one batch, any larger amount will require staff confirmation,
+# set this to 0 to require explicit approval for batches of any size
+KICK_CONFIRMATION_THRESHOLD = 0.01  # 1%
+
+BOT_MESSAGE_DELETE_DELAY = 10
```

Comment:
I'm not sure. I usually try to advocate keeping the definitions close to where they are used.

Let's imagine that these settings are in the `config.yml`. If you're reading the code and you wonder what the value is, you have to:
* Follow the import to `constants.py`
* Look at from where the value is being loaded
* Navigate to the YAML and find the value

IDE generally cannot aid you in this process, as the value only gets bound at runtime. In my opinion, this creates a significant level of indirection between the developer and the code. I think it's fair to say that these values have no use outside of this module, and so making them more accessible globally has no benefit.

On the other hand, having the constants here, your IDE can probably tell you what the value is if you just hover the reference, and if you want to change the values locally for debugging purposes, you can do that just as easily.

These are my thoughts on the topic, but I know that many would disagree. I wouldn't mind moving them to the config file, but I'd like to hear more of a justification for it. Because personally I cannot help but feel like I'm only making things more difficult for myself.

Reply:
Another downside that I notice with creating move distance between the value and its use, is that it increases the amount of necessary documentation. The names of these variables would likely need to be longer, and be more annotated, so that it's obvious what they represent.

Reply:
One additional factor to consider is if these values need to be changed for a testing environment. If so, using the config system ensures that people do not have the modify any of the files checked into our version control system, as they only have to configure `config.yaml` to their liking to overwrite the default in `config-default.yaml`. This prevents debug values from being accidentally committed to the repository.

Reply:
I'm honestly kind of torn about whether we should move it.  On one hand you're right, there's all the distance between the values and the and where they're used that causes potential issues sifting through the code, as I've certainly had in the past.  

I also think there's an argument to be made that the anti-spam cog needs it to be in the config file just due to the number of rules it touches.  But I think my main concern is when we actually need to tweak the configuration from a moderation standpoint.  Most of the other things that can be tweaked are in the config and are grouped together and easy enough to navigate just by using CTRL+F to find the right key word for the cog.  

If we do decide to keep them in the cog, I think we at the very least move any of the numerical constants to the top so that they're grouped together, as they're the ones we'd be more likely to need to tweak.   I'd be curious to hear the opinions of some of the others on this, though.  

And if nothing else it's something we can move at a later date should we decide to go that route.

Reply:
> This prevents debug values from being accidentally committed to the repository.

@SebastiaanZ: This is true, although it doesn't feel like a very strong argument to me. I don't think it's difficult to review the diff before commiting it, and it should definitely be done before pushing. Unintended diff being commited & pushed sadly happens often even with these measures in place. I actually probably find it easier to have them version controlled, it forces you to keep your settings tidy & you will always know when and how your configured behaviour differs from the production one.

With regard to debug values, keeping the values here will reload them if you reload the extension. This means that you don't have to restart the whole bot to tweak them. But that's more of a `d.py` detail. It is extremely useful when testing the feature though.

> But I think my main concern is when we actually need to tweak the configuration from a moderation standpoint.

I'd be perfectly fine with moving them if it makes things easier for others. For the reasons outlined in the paragraph above, I don't like working with the config yaml too much, but once the code hits production it will be a different situation too. From a maintenance perspective it may just make more sense.

I'm happy that we're talking about this though. Let's wait a little longer to see what others think.

Reply:
> This is true, although it doesn't feel like a very strong argument to me. I don't think it's difficult to review the diff before commiting it, and it should definitely be done before pushing. Unintended diff being commited & pushed sadly happens often even with these measures in place. I actually probably find it easier to have them version controlled, it forces you to keep your settings tidy & you will always know when and how your configured behaviour differs from the production one.

To me, personally, it's one of the biggest reasons why we use the configuration system we do. You can easily overwrite constants without having to touch files that actually have to be checked in. This means that you can keep your local debugging setup without having to revert values and us risking a debug value making it past review. I'd rather have a foolproof method than one that relies on human review, because we all make mistakes during reviews and sometimes features get merged with fairly minimal reviews.

I find the opposite approach tidy: You don't have to mess with checked in files, you don't have to be careful with committing changes to avoid committing lines then need to be reverted to keep a clean blame/commit history, you can persist your local debug setup independent of check in files.

Reply:
Ideally these would be configurable. It's just that the current config system sucks and we haven't gotten far with coming up with an alternative.

Reply:
I think for now, we should align with the current configuration standard and move the things we would need to tweak on the fly (the constants I mentioned up above) to the config-default.yaml

Reply:
Moved constants to the YAML file in 1b3a5c5 including `BOT_MESSAGE_DELETE_DELAY` which was an in-module constant before this PR. I'm not sure, but it doesn't look like the top-level definitions in neither `constants.py` nor the YAML are ordered lexicographically, so I just put mine below `PythonNews` in both.